src/rgw/rgw_lc_tier.cc

```diff
@@ -0,0 +1,1186 @@
+        return -EIO;
+      }
+    } else {
+      ldout(cct, 0) << "WARNING: acl attrs not provided" << dendl;
+    }
+    return 0;
+  }
+
+  int read(off_t ofs, off_t end, bufferlist &bl) {
+    optional_yield y = null_yield;
+
+    return read_op.read(ofs, end, bl, y);
+  }
+};
+
+
+class RGWLCStreamPutCRF : public RGWStreamWriteHTTPResourceCRF
+{
+  CephContext *cct;
+  RGWHTTPManager *http_manager;
+  rgw_lc_obj_properties obj_properties;
+  std::shared_ptr<RGWRESTConn> conn;
+  rgw::sal::RGWObject* dest_obj;
+  string etag;
+
+  public:
+  RGWLCStreamPutCRF(CephContext *_cct,
+      RGWCoroutinesEnv *_env,
+      RGWCoroutine *_caller,
+      RGWHTTPManager *_http_manager,
+      const rgw_lc_obj_properties&  _obj_properties,
+      std::shared_ptr<RGWRESTConn> _conn,
+      rgw::sal::RGWObject* _dest_obj) :
+    RGWStreamWriteHTTPResourceCRF(_cct, _env, _caller, _http_manager),
+    cct(_cct), http_manager(_http_manager), obj_properties(_obj_properties), conn(_conn), dest_obj(_dest_obj) {
+    }
+
+
+  int init() override {
+    /* init output connection */
+    RGWRESTStreamS3PutObj *out_req{nullptr};
+
+    if (multipart.is_multipart) {
+      char buf[32];
+      snprintf(buf, sizeof(buf), "%d", multipart.part_num);
+      rgw_http_param_pair params[] = { { "uploadId", multipart.upload_id.c_str() },
+        { "partNumber", buf },
+        { nullptr, nullptr } };
+      conn->put_obj_send_init(dest_obj, params, &out_req);
+    } else {
+      conn->put_obj_send_init(dest_obj, nullptr, &out_req);
+    }
+
+    set_req(out_req);
+
+    return RGWStreamWriteHTTPResourceCRF::init();
+  }
+
+  static bool keep_attr(const string& h) {
+    return (keep_headers.find(h) != keep_headers.end() ||
+        boost::algorithm::starts_with(h, "X_AMZ_"));
+  }
+
+  static void init_send_attrs(CephContext *cct, const rgw_rest_obj& rest_obj,
+                              const rgw_lc_obj_properties& obj_properties,
+                              map<string, string> *attrs) {
+
+    map<string, RGWTierACLMapping>& acl_mappings(obj_properties.target_acl_mappings);
+    string target_storage_class = obj_properties.target_storage_class;
+
+    auto& new_attrs = *attrs;
+
+    new_attrs.clear();
+
+    for (auto& hi : rest_obj.attrs) {
+      if (keep_attr(hi.first)) {
+        new_attrs.insert(hi);
+      }
+    }
+
+    auto acl = rest_obj.acls.get_acl();
+
+    map<int, vector<string> > access_map;
+
+    if (!acl_mappings.empty()) {
+      for (auto& grant : acl.get_grant_map()) {
+        auto& orig_grantee = grant.first;
+        auto& perm = grant.second;
+
+        string grantee;
+
+        const auto& am = acl_mappings;
+
+        auto iter = am.find(orig_grantee);
+        if (iter == am.end()) {
+          ldout(cct, 20) << "acl_mappings: Could not find " << orig_grantee << " .. ignoring" << dendl;
+          continue;
+        }
+
+        grantee = iter->second.dest_id;
+
```

Comment:
I see some partial similarities with the code at cloud sync, what prevented you from reusing same code? Maybe there's an opening for consolidation?

Reply:
@yehudasa  ..yes..along with config, most of the coroutine processing code is also similar. At the beginning of this project, I had bought this up in one of the refactoring call that if we can move most of the cloud sync code to a common place..but you had suggested that it shall become a lot more work than this feature actually is and you may be already working on refactoring that code. Sorry if I have misunderstood that part. 
If you suggest, I can work on consolidating both the modules code, atleast config part now.

Reply:
Let's try to strike a balance (easy to see path to consolidation, result not likely to be brittle, not too much longer to completion of this phase...)?

Reply:
have marked it as TODO - both code consolidation of the cloud-sync & cloud-transition modules and updating them with Zipper APIs. Will work on it once these changes get merged as it helps with testing too.